functorch/csrc/BatchingRegistrations.cpp

```diff
@@ -557,6 +557,32 @@ Tensor cat_batching_rule(TensorList tensors, int64_t dim) {
   return physical_views[0].getPhysicalToLogicalMap().apply(result);
 }
 
+Tensor block_diag_batching_rule(TensorList tensors) {
+  if (!participatesInCurrentLevel(tensors)) {
+    c10::impl::ExcludeDispatchKeyGuard guard(kBatchedKey);
+    return at::block_diag(tensors);
+  }
+  auto physical_views = MultiBatchVmapTransform::logicalToPhysical(tensors);
+  auto physical_tensors = fmap(
+      physical_views, [](const VmapPhysicalView& view) -> Tensor { return view.tensor(); });
+  TORCH_INTERNAL_ASSERT(
+      tensors.size() > 0, "The dispatcher should not have dispatched here otherwise.");
+  // Implementing this as a dummy for loop for now, since I'm not sure how to do it any better.
+  // I'm probably not accounting for potentially multiple batched dimensions?
+  auto bdim = tensors[0].size(0);
+  std::vector<Tensor> batched_outputs;
+  for (const auto& i : c10::irange(bdim)) {
+    std::vector<Tensor> inputs_for_batch;
+    for (const auto& t : physical_tensors) {
+      inputs_for_batch.push_back(t[i]);
+    }
+    auto out_for_batch = at::block_diag(inputs_for_batch);
+    batched_outputs.push_back(out_for_batch);
+  }
+  auto result = at::cat(batched_outputs);
+  return physical_views[0].getPhysicalToLogicalMap().apply(result);
+}
```

Comment:
I'm not convinced that I actually implemented this correctly (particularly in the case where you have multiple layers of vmap), but I wasn't sure what the right API's to use were.

I also wasn't sure how to actually make this batching rule fast, so I implemented it as a dummy for loop. It's probably still more efficient than the FUNCTIONAL_DECOMPOSE version though, since that functionalizes every intermediate `copy_()` which probably resulted in a bunch of large temporary tensors.